etcdserver/snapshot_store.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	dstorage "github.com/coreos/etcd/storage"
+)
+
+type snapshot struct {
+	r raftpb.Snapshot
+}
+
+func (s *snapshot) raft() raftpb.Snapshot { return s.r }
+
+func (s *snapshot) size() uint64 { return 0 }
+
+func (s *snapshot) writeTo(w io.Writer) (n int64, err error) { return 0, nil }
+
+func (s *snapshot) close() {}
+
+type snapshotStore struct {
+	// dir to save snapshot data
+	dir string
+	kv  dstorage.KV
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// snapshotStore will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan raftpb.Snapshot
+
+	mu   sync.Mutex
+	snap *snapshot
+}
+
+func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
+	return &snapshotStore{
+		dir:      dir,
+		kv:       kv,
+		reqsnapc: make(chan struct{}),
+		snapc:    make(chan raftpb.Snapshot),
+	}
+}
+
+// getNew gets a new snapshot.
+// It returns ErrSnapshotTemporarilyUnavailable if it fails to return snapshot now.
+func (ss *snapshotStore) getNew() (*snapshot, error) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	if ss.snap != nil {
+		return nil, raft.ErrSnapshotTemporarilyUnavailable
+	}
+
+	// ask to generate v2 snapshot
+	ss.reqsnapc <- struct{}{}
+	// TODO: generate v3 snapshot at here
+	raftsnap := <-ss.snapc
+	ss.snap = &snapshot{
+		r: raftsnap,
+	}
+	return ss.snap, nil
+}
+
+// get gets the generated snapshot.
+func (ss *snapshotStore) get() *snapshot {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	return ss.snap
+}
+
+// del deletes the generated snapshot.
+func (ss *snapshotStore) del() {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	ss.snap.close()
+	ss.snap = nil
+}
+
+// save saves snapshot into disk.
+//
+// If snapshot has existed, it returns error and keeps the original snapshot
+// data. The function guarantees that if the call is aborted at any time,
```

Comment:
what do you mean by abort? it cannot be aborted. This method saves snapshot to disk atomically, but might leave broken snapshot file with tmp suffix when fails.